src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1895 @@
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(Job*, job_free);
+DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR(job_hash_ops, uint64_t, uint64_hash_func, uint64_compare_func,
+                                      Job, job_free);
+
+static int job_new(JobType type, Target *t, sd_bus_message *msg, JobComplete cb, Job **ret) {
+        _cleanup_(job_freep) Job *j = NULL;
+        int r;
+
+        assert(t);
+        assert(ret);
+
+        j = new(Job, 1);
+        if (!j)
+                return -ENOMEM;
+
+        *j = (Job) {
+                .type = type,
+                .target = t,
+                .manager = t->manager,
+                .id = t->manager->current_job_id + 1,
+                .stdout_fd = -EBADF,
+                .complete_cb = cb,
+                .dbus_msg = sd_bus_message_ref(msg),
+        };
+
+        if (asprintf(&j->object_path, "/org/freedesktop/sysupdate1/job/_%" PRIu64, j->id) < 0)
+                return -ENOMEM;
+
+        r = hashmap_ensure_put(&t->manager->jobs, &job_hash_ops, &j->id, j);
+        if (r < 0)
+                return r;
+
+        t->manager->current_job_id = j->id;
+
+        *ret = TAKE_PTR(j);
+        return 0;
+}
+
+static int job_parse_child_output(int _fd, sd_json_variant **ret) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_close_ int fd = _fd;
+        _cleanup_fclose_ FILE *f = NULL;
+        struct stat st;
+        int r;
+
+        assert(ret);
+
+        if (fstat(fd, &st) < 0)
+                return log_error_errno(errno, "Failed to stat stdout fd: %m");
+
+        assert(S_ISREG(st.st_mode));
+
+        if (st.st_size == 0)
+                return 0;
+
+        if (lseek(fd, SEEK_SET, 0) == (off_t) -1)
+                return log_error_errno(errno, "Failed to seek to beginning of memfd: %m");
+
+        f = take_fdopen(&fd, "r");
+        if (!f)
+                return log_error_errno(errno, "Failed to reopen memfd: %m");
+
+        r = sd_json_parse_file(f, "stdout", 0, &v, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+
+static void job_on_ready(Job *j) {
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *msg = NULL;
+        int r;
+
+        assert(j);
+
+        /* Some jobs run in the background as we return the job ID to the
+         * dbus caller (i.e. for the Update method). However, the worker
+         * will perform some sanity-checks on startup which would be valuable
+         * as dbus errors. So, we wait for the worker to signal via READY=1
+         * that it has completed its sanity-checks and we should continue
+         * the job in the background. */
+
+        if (!j->detach_on_ready)
+                return;
+
+        assert(j->dbus_msg);
+        msg = TAKE_PTR(j->dbus_msg);
+
+        j->complete_cb = NULL;
+
+        r = sd_bus_reply_method_return(msg, "suo", j->version, j->id, j->object_path);
```

Comment:
hmm "u" expects 32bit. but j->id is 64bit. this will result in a bad memory access on BE archs

I think we should just make the job ids 64bit, it's not worth the trouble to have too short ids. hence please switch this the id to 64bit also externally, i.e. in the entire dbus interface.